### ISSUE TITLE
Nested optional types encoded with zero length

### ISSUE THREAD

**Issue by vanrein** [vanrein/quick-der#13]
This applies to TLS-KDH, just preceding the checkin of https://github.com/arpa2/tlspool/commit/c78ab027474902182d63ed5d1a13832ea5423099

In Kerberos' `Authenticator`:

```
   Authenticator   ::= [APPLICATION 2] SEQUENCE  {
           authenticator-vno       [0] INTEGER (5),
           crealm                  [1] Realm,
           cname                   [2] PrincipalName,
           cksum                   [3] Checksum OPTIONAL,
           cusec                   [4] Microseconds,
           ctime                   [5] KerberosTime,
           subkey                  [6] EncryptionKey OPTIONAL,
           seq-number              [7] UInt32 OPTIONAL,
           authorization-data      [8] AuthorizationData OPTIONAL
   }
```

The parts `[7] UInt32 OPTIONAL` with no value of that kind in the corresponding `dercursor`, run through `der_pack()`, leads to the encoding `a7 00`.  Similarly `[8] AuthorizationData OPTIONAL` is packed as `a8 00`.  The result is unacceptable to `der_unpack()`.

In other words, the `OPTIONAL` is not applied to the outside tag and only to the inner real type.  This is wrong.

**Comment by vanrein** [comment]
Here's an example dump of an `Authenticator` from `der_pack()`:

-   a0 03 02 01 05
-   a1 0b 1b 09 41 52 50 41 32 2e 4e 45 54
-   a2 24 30 22
    -   a0 03 02 01 0a
    -   a1 1b 30 19 74 65 73 74 63 6c 69 40 74 6c 73 70 6f 6f 6c 2e 61 72 70 61 32 2e 6c 61 62
-   a3 09 30 07
    -   a0 03 02 01 fe — checksumtype 254
    -   a1 00 — no checksum bytes? — why is this included with length 00?
-   a4 05 02 03 00 da 68
-   a5 1a 18 18 54 68 75 20 41 75 67 20 32 35 20 31 35 3a 30 35 3a 33 31 20 32 30 31 36
-   a6 2b 30 29 a0 03 02 01 12 a1 22 04 20 4e df b0 f9 a4 1f 34 cf e1 85 cb 0b 7c d6 a3 d6 83 0e 4b 08 cc 93 20 b1 6a f3 74 0d e7 b2 c6 d5
-   a7 00 — why is this included with length 00?
-   a8 00 — why is this included with length 00?

In this, 

 * WILD & WEIRD, INCLUSION OF ABSENT OPTIONALS’ TAGS WITH ZERO LENGTH
 * SOLVED: BAD & BRAVEN, CHECKSUM TYPE BUT NO CHECKSUM BYTES INCLUDED

**Comment by vanrein** [comment]
The checksum has now been included in commit c78ab027474902182d63ed5d1a13832ea5423099

**Comment by vanrein** [comment]
Issue solved in commit https://github.com/vanrein/quick-der/commit/ccdcf13be1584652e9c2de9fdf180a187b64dc86
This closes the case, thanks for reporting :-D

**Comment by vanrein** [comment]
Additionally, there was a problem parsing these structures (that _end_ in OPTIONAL ASN.1 tags, even nested tags at times).  This has been resolved in https://github.com/vanrein/quick-der/commit/7ff61acaf2718592f31bd1e7d077cc537701e6e8
